### PR TITLE
Simplify content file globbing and remove FileProviders.Abstractions and Primitives from VSIX

### LINKED ISSUE #12696: Perf: ContentFileUtils should cache/reuse virtual file instances
### NuGet Product Used

dotnet.exe, MSBuild.exe, NuGet.exe

### Product Version

all

### Worked before?

_No response_

### Impact

None

### Repro Steps & Context

This codepath is used when restoring PackageReference projects with packages that have `contentFiles/`. Therefore, restore performance could be improved.

Note, this will be less beneficial if the following PR and issues are fixed:
* https://github.com/NuGet/NuGet.Client/pull/5231
* https://github.com/NuGet/Home/issues/12695

Currently, when ContentFileUtils is trying to [match `<contentFiles>` entries in the nuspec to `contentFiles/` files in the zip](https://github.com/NuGet/NuGet.Client/blob/4b6a87bfe90d247f961eba41d63dbf418a56fdcc/src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs#L104-L148), it has a nested loop, and creates a new `VirtualFileProvider` (will soon change to `SingleFileProvider`) within the inner loop. This means when there are 10 patterns in the nuspec, each file in the zip will have a provider created for it 10 times. If there are 20 files, then that's 200 provider objects created to match these 20 files.

Using [NuGet Insights](https://github.com/NuGet/Insights), I see that of packages that have any `<contentFiles>` with a `*` in their nuspec, the average is 1.2 entries and the max is 21 entries. However, only 143 packages contain more than 1 `<contentFiles>` pattern in their nuspec, and of those only 3 packages have over 1 million downloads. Although one of them is marked is marked as deprecated. So, overall, it's hard to say whether implementing this will actually have a meaningful impact. The `SingleFileProvider` might actually be better for most packages, since few packages use file globbing with more than 1 pattern.

### Verbose Logs

_No response_

### LINKED ISSUE #14686: Investigate removing Microsoft.Extensions.* assemblies from the VSIX
I removed `Microsoft.Extensions.DependencyInjection.*` assemblies from the VSIX in https://github.com/NuGet/NuGet.Client/pull/7007, but it may be possible to remove these as well:
 
- Microsoft.Extensions.FileProviders.Abstractions.dll
- Microsoft.Extensions.FileSystemGlobbing.dll
- Microsoft.Extensions.Primitives.dll

MSBuild already has `Microsoft.Extensions.Primitives.dll` in its bin folder, so that one should be good.
See what actually uses the other assemblies and remove them from the VSIX if appropriate.

### PR DESCRIPTION
**Pull request by donnie-msft**
# Bug

 Fixes: https://github.com/NuGet/Home/issues/14686
 Fixes: https://github.com/NuGet/Home/issues/12696
 
  ## Description
  
  Removes `Microsoft.Extensions.FileProviders.Abstractions.dll` and `Microsoft.Extensions.Primitives.dll` from the VSIX by eliminating the code dependency on them.
  
  ### What changed
  
- **Replaced** the custom virtual filesystem adapter (`SingleFileProvider`, `FileProviderGlobbingDirectory`, 
`FileProviderGlobbingFile`, `VirtualFileInfo`) with a direct call to `matcher.Match("ROOT", relativePath)`. No custom classes needed.
 - **Deleted** 4 source files (~370 lines) — net reduction of ~370 lines.
 - **Removed** `Microsoft.Extensions.FileProviders.Abstractions` package reference from `NuGet.Commands.csproj` 
and `Directory.Packages.props`.
 - **Updated** `.vsixinclude`, `.vsixignore`, `.swr`, and `ilmerge.props` to exclude the two removed assemblies.
  
  ### What stays
  
  - `Microsoft.Extensions.FileSystemGlobbing` remains — it provides the `Matcher` class used for content file glob pattern 
matching.
 
  ### Performance
 
 Benchmarked the content file globbing hot path across 2,428 packages from the global packages folder (20,540 match operations per iteration):
 
 | Approach | Mean | Allocated |
 |----------|------|-----------|
 | Old code (4 classes + IFileProvider) | 42.66 ms | 84.21 MB |
 | SingleFileDirectory Approach | 32.94 ms | 64.20 MB |
 | **New code — `Match("ROOT", path)`** | **28.61 ms** | **45.46 MB** | 

The new approach is **~25% faster** and allocates **~46% less memory** by using `Matcher.Match("ROOT", 
relativePath)` directly, avoiding `IFileProvider` indirection and virtual file info object creation entirely.
 
 ### Validation
  
 Validated on a clean VM (VS installed from Apex & E2E pipeline bootstrapper, MSBuild 18.7.0-preview) that the refactored 
content file globbing produces identical restore output.
  
 - Created a test project referencing 12 packages with content files spanning a variety of nuspec globbing patterns (recursive wildcards, backslash separators, framework-scoped paths, include+exclude combinations, etc.).
 - Compared the `contentFiles` section of `project.assets.json` between stock SDK NuGet and this branch.
 - **Results** (831 content file entries compared):
    - ✅ `msbuild -t:Restore` (branch CLI) — identical to baseline
    - ✅ VS auto-restore (branch VSIX) — identical to baseline
## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests - No - instead I manually validated with a bunch of different Contentfile Include/Exclude patterns as described above.
- [ ] Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.

### CODE REVIEW

**Review by zivkan — DISMISSED** [review]
FWIW, I'm not a fan of the AI generated PR description. It duplicated some information, and gave some information that I consider not helpful as a reviewer. I feel like it kind of misses the point, or at least put me on slightly the wrong mental context, when thinking about the manual testing that happened on the VM (the focus was to test the removed assemblies didn't cause failures on msbuild restore, not to do exhaustive edge case testing, which really should be done with unit tests, not manual tests).

**Comment by donnie-msft** [comment]
Thanks for the feedback, @zivkan ! Updated the Validation section in the PR Description to be more focused on the validation of restore, not necessarily edge cases for content file patterns.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs [hunk 1]
```diff
@@ -130,16 +127,7 @@ internal static List<LockFileContentFile> GetContentFileGroup(
...
-                            fileProvider: new SingleFileProvider(relativePath),
-                            fileInfo: rootDirectory,
-                            parent: null);
-
-                        // Currently Matcher only returns the file name not the full path, each file must be
-                        // check individually.
-                        var matchResults = matcher.Execute(globbingDirectory);
-
-                        if (matchResults.HasMatches)
+                        if (matcher.Match(relativePath).HasMatches)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: Could that fact that we're not passing a root directory here cause issues?

Basically the method internally seems to call Directory.GetCurrentDirectory, while before we would just do root. 

I just don't want to accidentally hit max path issues. It probably makes it more performant by having a shorter path to combine to.

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: Good catch. addressed by switching from `Matcher.Match(relativePath)` to `Matcher.Execute(new SingleFileDirectory(relativePath))`. 

Added some benchmark results to the PR Description. Looks like this may be more efficient than what we originalily had.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: That's great to hear that we're doing even better now!

My suggestion was actually just calling `matcher.Match("ROOT", relativePath)`. 

If you still have the benchmarks, can you try that quicky and see if it makes a difference?

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: Your suggestion appears to be the most efficient, and it's now in this PR!

I used copilot to rewrite a benchmark and compare scenarios. See the commit history for the full benchmark (now removed in latest commit).

| Approach                          | Mean     | Allocated |
|---------------------------------- |---------:|----------:|
| Branch_MatchRelativePath (old)    | 73.41 ms | 135.87 MB |
| Dev_ExecuteWithFileProvider       | 38.37 ms |  84.21 MB |
| Proposed_Split (SingleFileDir)    | 31.21 ms |  64.20 MB |
| MatchRoot_RelativePath            | 28.61 ms |  45.46 MB |
| Proposed_Span                     | 35.07 ms |  70.78 MB |

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: That's awesome!

**Review by nkolev92 — DISMISSED** [review]

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: unfortunately there's some test failures.

**Comment by donnie-msft** [comment]
@copilot some tests failed. Can you investigate this error and attempt a fix in this PR?: 

Error message
Assert.True() Failure
Expected: True
Actual:   False
Stack trace
   at NuGet.Commands.Test.ContentFilesTests.<ContentFiles_NuspecContentFilesGlobbingSupportExcludeDirSet>d__13.MoveNext()

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: Looks like copilot reactored it to fix the tests. Have a look and see what you think. If those tests past, seems to be we skills be good.

**Review by zivkan — APPROVED** [review]
I ❤️ perf.  🚗 ➡️ 🚀

**Review by jeffkl — APPROVED** [review]

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs]: Followed-up with Nikolche offline to let him know the tests were resolved.
